example/socks/socks_client.hpp

```diff
@@ -0,0 +1,491 @@
+        };
+        enum {
+            SOCKS_CMD_CONNECT = 0x01,
+            SOCKS_CMD_BIND = 0x02,
+            SOCKS5_CMD_UDP = 0x03
+        };
+        enum {
+            SOCKS5_ATYP_IPV4 = 0x01,
+            SOCKS5_ATYP_DOMAINNAME = 0x03,
+            SOCKS5_ATYP_IPV6 = 0x04
+        };
+        enum {
+            SOCKS5_SUCCEEDED = 0x00,
+            SOCKS5_GENERAL_SOCKS_SERVER_FAILURE,
+            SOCKS5_CONNECTION_NOT_ALLOWED_BY_RULESET,
+            SOCKS5_NETWORK_UNREACHABLE,
+            SOCKS5_CONNECTION_REFUSED,
+            SOCKS5_TTL_EXPIRED,
+            SOCKS5_COMMAND_NOT_SUPPORTED,
+            SOCKS5_ADDRESS_TYPE_NOT_SUPPORTED,
+            SOCKS5_UNASSIGNED
+        };
+        enum {
+            SOCKS4_REQUEST_GRANTED = 90,
+            SOCKS4_REQUEST_REJECTED_OR_FAILED,
+            SOCKS4_CANNOT_CONNECT_TARGET_SERVER,
+            SOCKS4_REQUEST_REJECTED_USER_NO_ALLOW,
+        };
+
+    public:
+        explicit socks_client(tcp::socket& socket,
+            const std::string& hostname, unsigned short port, int version,
+            const std::string& username, const std::string& password,
+            bool use_hostname, Handler handler)
+            : socket_(socket)
+            , hostname_(hostname)
+            , port_(port)
+            , version_(version)
+            , username_(username)
+            , password_(password)
+            , use_hostname_(use_hostname)
+            , handler_(handler)
+        {}
+
+        void handshake()
+        {
+            auto self = std::enable_shared_from_this<socks_client<Handler>>::shared_from_this();
+            switch (version_)
+            {
+            case SOCKS_VERSION_5:
+                boost::asio::spawn(socket_.get_io_context(), [this, self]
+                (boost::asio::yield_context yield)
+                {
+                    do_socks5_handshake(yield);
+                });
+            break;
+            case SOCKS_VERSION_4:
+                boost::asio::spawn(socket_.get_io_context(), [this, self]
+                (boost::asio::yield_context yield)
+                {
+                    do_socks4_handshake(yield);
+                });
+            break;
+            default:
+                BOOST_ASSERT("socks version error!" && false);
+            }
+        }
+
+    private:
+        template<typename type, typename source>
+        type read(source& p)
+        {
+            type ret = 0;
+            for (std::size_t i = 0; i < sizeof(type); i++)
+                ret = (ret << 8) | (static_cast<unsigned char>(*p++));
+            return ret;
+        }
+
+        template<typename type, typename target>
+        void write(type v, target& p)
+        {
+            for (auto i = (int)sizeof(type) - 1; i >= 0; i--, p++)
+                *p = static_cast<unsigned char>((v >> (i * 8)) & 0xff);
+        }
+
+        void do_socks5_handshake(boost::asio::yield_context yield)
+        {
+            boost::asio::streambuf request;
```

Comment:
instead of `asio::streambuf`, consider using `beast::static_buffer` or `beast::flat_static_buffer`. Rationale: If the strings which make up the SOCKS handshake have an upper limit on their size, use of a static buffer can eliminate the need for a dynamic memory allocation. If you do this, you may need to also use `handler_ptr` since a composed operation state containing a static buffer is non-trivial.